pipenv/utils/resolver.py

```diff
@@ -552,6 +560,28 @@ def constraint_file(self):
             self._constraint_file = self.prepare_constraint_file()
         return self._constraint_file
 
+    def prepare_default_constraint_file(self):
+
+        from pipenv.vendor.vistir.path import create_tracked_tempfile
+
+        default_constraints_file = create_tracked_tempfile(
+            mode="w",
+            prefix="pipenv-",
+            suffix="-default-constraints.txt",
+            dir=self.req_dir,
+            delete=False,
+        )
+        default_constraints = get_constraints_from_deps(self.project.packages)
+        default_constraints_file.write("\n".join([c for c in default_constraints]))
+        default_constraints_file.close()
+        return default_constraints_file.name
+
+    @property
```

Comment:
Can we make this a `@cached_property` instead so that we don't have to implement that ourselves with `if self._default_constraint_file is None:` logic?

Reply:
I used ``@cached_property`` but falling some tests on MacOS.

Reply:
Those tests also failed without ``@cached_property`` as well, so maybe ``@cached_property`` is not the issue. Because ``@cached_property`` does not work with python 3.7, I used ``@property`` and ``@lru_cache`` instead.

Reply:
I think the problem is you need to import `@cached_property` from the vendor'd dependencies, go with `from pipenv.vendor.cached_property import cached_property` until https://github.com/pypa/pipenv/issues/5169 is completed.

Reply:
The only tests that are expected could fail, and likely on mac OS right now is `test_convert_deps_to_pip` and `test_convert_deps_to_pip_one_way`.  Still investigating how these failures started in the main branch, so if you see that specifically its safe to ignore.  The other tests should all pass.

Reply:
> I think the problem is you need to import `@cached_property` from the vendor'd dependencies, go with `from pipenv.vendor.cached_property import cached_property` until #5169 is completed.

Thank you, I'll try that.

Reply:
I ran with the main branch, but there are more failed tests on macOS other than ``test_convert_deps_to_pip`` and ``test_convert_deps_to_pip_one_way``.

Reply:
@dqkqd It is possible that the new version of setuptools that was released yesterday has affected things.  I will re-run the  main branch actions, but you can see the last time it ran upon merge ony those `*_depts_to_pip` tests had failed:  https://github.com/pypa/pipenv/runs/7758310446?check_suite_focus=true

Reply:
@dqkqd I fixed the unexpected test failures upstream in main branch, in this PR:  https://github.com/pypa/pipenv/pull/5241

I am still not sure why those *_depts_to_pip tests have sporadic failures in the CI and Mac OS, but thats another can of worms.

Reply:
@matteius I updated as you said. Not sure why it passed ``*_depts_to_pip`` cases. I didn't even modify ``convert_deps_to_pip``.

Reply:
@dqkqd It is a "flakey" test case, maybe I'll mark it as such and it won't fail the build as often.  I haven't gotten to the bottom of why sometimes it behaves differently, it has failed on Mac OS most frequently, but I have seen it once in a while fail on other systems.